File: Union Find/Quick Find/QuickFind.java
package com.company;

public class QuickFIndUF {
    private int id[];
    public QuickFIndUF(int N){
        id = new int[N];
        for(int i = 0;i<N;i++){
            id[i] = i;
        }
    }
    public boolean connected(int p,int q){
        return id[p] == id[q];
    }
    public void union(int p, int q){
        int pid = id[p];
        int qid = id[q];
        for(int i = 0;i<id.length;i++){
            if(id[i] == pid){
                id[i]= qid;
            }
        }
    }
    public static void main(String args[]){
        QuickFIndUF a = new QuickFIndUF(10);
        a.union(1,2);
        a.union(2,3);
        System.out.println(a.connected(1,3));
    }
}

//Quick Find defect - Union is too expensive ( quadratic time complexity)

File: Union Find/Quick Find/QuickFind.c
//Quick Find
#include<stdio.h>
#include<stdlib.h>
int *id = NULL;
int n = 0;
int connected(int p,int q){
    if(id[p] == id[q])
    {
        return 1;
    }
    return 0;
}

void Union(int p,int q){
    int pid,qid;
    pid = *(id+p);
    qid = *(id+q);
    for(int i= 0;i < n;i++){
        if(id[i] == pid)
            id[i] = qid;
    }
}

int main(){
    
    printf("Enter the value of n\n");
    scanf("%d",&n);
    id = (int*)malloc(sizeof(int)*n);
    for(int i = 0;i<n;i++){
        id[i] = i;
    }
    Union(1,2);
    
    if(connected(1,2))
        printf("True");
    else
        printf("False");
    return 0;
}
